WIP DNM esp32: add flash cache support

GautierAtWork opened the pull request:
Needs option ```CONFIG_BOOTLOADER_ESP_IDF```

Closes #3722 

todo:
- [x]  Add option in Kconfig
- [x] Add ```#ifdef``` around  changes in ```pinmux_esp32.c```

@lpereira: an option like ```CONFIG_ESP32_FLASH_CACHE``` will be introduced. Two linker scripts will exist then: one with data cache, one without data cache. Will it be ok for you?

GautierAtWork committed:
esp32: add flash cache support

Add flash cache support for esp32.

Signed-off-by: Gautier Seidel <gautier.seidel@tado.com>

lpereira reviewed (changes requested):
Nice catch on the pinmux code!

I think it's fine if we have two linker scripts, at least short term.  Long term, it should be better to only support building with flash cache enabled; otherwise, it's kind of hard to utilize the full potential of this SoC.

Patch LGTM, but I have a few questions.

andyross commented on drivers/pinmux/pinmux_esp32.c:
```diff
@@ -164,7 +164,12 @@ static int pinmux_initialize(struct device *device)
-		pinmux_set(NULL, pin, 0);
+		/* pin 6 to 11 are configured for SPI by the second stage bootloader.
+		 * We should not change them, otherwise we can't access the flash
+		 * anymore. */
```
If the point here is to disallow software from ever touching those pins through this driver, you probably want to put that rule into pinmux_set(), no?  

Honestly, I'm not sure I see much value to this initialization loop anyway.  If we have code enabled that cares about particular pin state, it will set it in its own initialization.  If not, it's probably better to rely on whatever the bootloader set for us.  Forcing "function zero" to be the default for everyone always is just going to cause collisions like this.  Maybe just remove the loop?

GautierAtWork replied:
I updated the comment. The idea is to keep the ability to reconfigure these pins during run-time.

gautierseidel committed:
esp32: add Kconfig flash cache option

Added in board Kconfig for esp32

Signed-off-by: Gautier Seidel <gautier@seidel.fr>

gautierseidel committed:
esp32: pinmux spi pins whitelist

esp32 pinmux driver doesn't re-initialize the SPI pins if flash cache
is used.

Signed-off-by: Gautier Seidel <gautier@seidel.fr>

gautierseidel committed:
esp32: add linker script with flash cache support

Now two linker scripts are available: one with flash cache suport,
one without.

Signed-off-by: Gautier Seidel <gautier@seidel.fr>

aescolar commented:
@GautierAtWork This PR seems to have been getting stale for very long.
It would seem you were waiting for some re-reviews(?) If so could you ping those?
Otherwise if you are not planning to continue working on this would you want to close it?
